File: src/com/chbachman/cminus/representation/statement/StructConstructor.java
package com.chbachman.cminus.representation.statement;

import com.chbachman.cminus.CMinusParser;
import com.chbachman.cminus.representation.Scope;
import com.chbachman.cminus.representation.Struct;
import com.chbachman.cminus.representation.function.Function;
import com.chbachman.cminus.representation.value.Value;
import com.chbachman.cminus.representation.value.Variable;

import java.util.List;
import java.util.Optional;
import java.util.stream.Collectors;

/**
 * Created by Chandler on 4/23/17.
 */
public class StructConstructor extends FunctionCall {
    public final Struct struct;

    public StructConstructor(CMinusParser.FunctionCallContext ctx, Scope scope) {
        super(ctx, scope, makeFunction(ctx, scope));

        // Since we already checked all this in the 'makeFunction' call, we can just directly access it.
        this.struct = scope.getStruct(ctx.ID().getText()).get();
    }

    private static Function makeFunction(CMinusParser.FunctionCallContext ctx, Scope scope) {
        // TODO: Fix this
        String name = ctx.ID().getText();
        Optional<Struct> struct = scope.getStruct(name);
        List<Value> parameters = ctx.value().stream().map(v -> Value.parse(v, scope)).collect(Collectors.toList());
        Function ref;

        if (struct.isPresent()) {
            ref = struct.get().getInit(parameters).get();
        } else {
            throw new RuntimeException("Function " + name + " was not found.");
        }

        return ref;
    }
}

File: src/main/java/com/chbachman/cminus/representation/statement/StructConstructor.kt
package com.chbachman.cminus.representation.statement

import com.chbachman.cminus.gen.CMinusParser
import com.chbachman.cminus.representation.Parser
import com.chbachman.cminus.representation.Scope
import com.chbachman.cminus.representation.Type
import com.chbachman.cminus.representation.value.Identifier

/**
 * Created by Chandler on 4/23/17.
 * Handles the construction of Structs
 */
class StructConstructor(ctx: CMinusParser.FunctionCallContext, scope: Scope) : Call {
    override val name = Identifier(ctx, scope)
    override val parameters = Parser.parse(ctx.funcSegment().argumentList(), scope)
    val struct = Type.getStruct(name.text)?.getStruct(scope)
            ?: throw RuntimeException("Could not find function")

    override val ref = struct.inits.map { it.header }.find { it.matches("init" + name, parameters) }
            ?: throw RuntimeException("Parameters didn't match")

    init {
        verify()
    }
}

